ktor-client/ktor-client-features/ktor-client-json/ktor-client-serialization/common/src/io/ktor/client/features/json/serializer/KotlinxSerializer.kt

```diff
@@ -92,5 +94,25 @@ private fun buildSerializer(value: Any): KSerializer<*> = when (value) {
 }
 
 @UseExperimental(ImplicitReflectionSerializer::class)
-private fun Collection<*>.elementSerializer(): KSerializer<*> =
-    firstOrNull()?.let { buildSerializer(it) } ?: String.serializer()
+private fun Collection<*>.elementSerializer(): KSerializer<*> {
+    val serializers = filterNotNull().map { buildSerializer(it) }.distinctBy { it.descriptor.name }
+
+    if (serializers.size > 1) {
+        error("Different serializers selected: ${serializers.map { it.descriptor.name }}")
```

Comment:
Not sure this is the most approachable error message.
Something like "Collection has heterogeneous elements that cannot be serialized by a single serializer"
Also, it is not clear how to fix this problem on the user's side